tutorial/lesson_21_auto_scheduler_generate.cpp

```diff
@@ -0,0 +1,195 @@
+        f(x, y, z, n) += filter(r.x, r.y, r.z, z) * input(x + r.x, y + r.y, r.z, n);
+        output1(x, y, z, n) = max(0, f(x, y, z, n));
+        output2(x, y, z, n) = min(min_value, f(x, y, z, n));
+    }
+
+    void schedule() {
+        if (auto_schedule) {
+            // To use the auto-scheduler, we need to provide estimates on all
+            // the input/output sizes including estimates on all the parameter
+            // values; otherwise, the auto-scheduler will throw an assertion.
+
+            // To provide estimates (min and extent values) for each dimension
+            // of the input images ('input', 'filter', and 'bias'), we use the
+            // set_bounds_estimate() method. set_bounds_estimate() takes in
+            // (min, extent) of the corresponding dimension as arguments.
+            input.dim(0).set_bounds_estimate(0, 131);
+            input.dim(1).set_bounds_estimate(0, 131);
+            input.dim(2).set_bounds_estimate(0, 64);
+            input.dim(3).set_bounds_estimate(0, 4);
+
+            filter.dim(0).set_bounds_estimate(0, 3);
+            filter.dim(1).set_bounds_estimate(0, 3);
+            filter.dim(2).set_bounds_estimate(0, 64);
+            filter.dim(3).set_bounds_estimate(0, 64);
+
+            bias.dim(0).set_bounds_estimate(0, 64);
+
+            // To provide estimates on the parameter values, we use the
+            // set_estimate() method.
+            min_value.set_estimate(2.0f);
+
+            // To provide estimates (min and extent values) for each dimension
+            // of pipeline outputs, we use the estimate() method. estimate()
+            // takes in (dim_name, min, extent) as arguments.
+            output1.estimate(x, 0, 128)
+                .estimate(y, 0, 128)
+                .estimate(z, 0, 64)
+                .estimate(n, 0, 4);
+
+            output2.estimate(x, 0, 128)
+                .estimate(y, 0, 128)
+                .estimate(z, 0, 64)
+                .estimate(n, 0, 4);
+
+            // Technically, the estimate values can be anything, but the closer
+            // they are to the actual use-case values, the better the generated
+            // schedule will be.
+
+            // Now, let's auto-schedule the pipeline by calling auto_schedule_outputs(),
+            // which takes in a MachineParams object as an argument. The machine_params
+            // argument is optional. If none is specified, the default machine parameters
+            // for a generic CPU architecture are going to be used by the auto-scheduler.
+
+            // Let's use some arbitrary but plausible values for the machine parameters.
+            MachineParams machine_params(32, 16 * 1024 * 1024, 40);
+            // The arguments to MachineParams are the maximum level of parallelism
+            // available, the size of the last-level cache (in KB), and the ratio
+            // between memory cost and arithmetic cost at the last level case, of
```

Comment:
between the cost of arithmetic and *the cost of a miss at the last-level cache*

Reply:
Would this be maybe more readable if named constants used? e.g.

```
const int kParallelism = 32;
const int kLastLevelCacheSize = 16 * 1024 * 1024;
const int kBalance = 40;
MachineParams machine_params(kParallelism, kLastLevelCacheSize, kBalance);
```